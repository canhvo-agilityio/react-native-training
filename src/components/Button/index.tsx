import { memo } from 'react';
import { ActivityIndicator, TouchableOpacity, Text } from 'react-native';

//Type
import { ButtonSizeTypes, ButtonVariantTypes } from '@/types';
import {
  buttonSizes,
  buttonTextSize,
  buttonTextStyle,
  buttonVariants,
  colors,
} from '@/themes';
export interface ButtonProps {
  text: string;
  disabled?: boolean;
  variant?: ButtonVariantTypes;
  size?: ButtonSizeTypes;
  isLoading?: boolean;
  onPress?: () => void;
}

const Button = ({
  text,
  variant = 'solid',
  size = 'md',
  disabled,
  isLoading,
  onPress,
}: ButtonProps) => {
  return (
    <TouchableOpacity
      style={[
        buttonSizes[size],
        buttonVariants[variant],
        (disabled || isLoading) && { opacity: 0.7 },
        { alignSelf: 'flex-start' },
      ]}
      onPress={onPress}
      disabled={disabled || isLoading}>
      <Text
        style={[
          buttonTextStyle.baseStyle,
          buttonTextStyle[variant],
          buttonTextSize[size],
        ]}>
        {text}
      </Text>
      {isLoading && (
        <ActivityIndicator size="small" color={colors.common.white} />
      )}
    </TouchableOpacity>
  );
};

export default memo(Button);
