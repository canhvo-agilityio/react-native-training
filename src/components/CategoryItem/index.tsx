import { CategoryItemType } from '@/types';
import { memo } from 'react';
import { Image, StyleSheet, TouchableOpacity, View } from 'react-native';
import Text from '../Text';
import { colors } from '@/themes';

interface CategoryItemProps {
  item: CategoryItemType;
  isActive: boolean;
  onPress: (id: string) => void;
}

const CategoryItem = ({ item, isActive, onPress }: CategoryItemProps) => (
  <TouchableOpacity
    testID="category"
    activeOpacity={1}
    style={styles.wrapper}
    onPress={() => onPress(item.id)}>
    <View style={{ ...styles.image, ...(isActive && styles.imageActive) }}>
      <Image source={item.image} alt={item.label} />
    </View>
    <Text
      color={isActive ? colors.background.primary : colors.border.quinary}
      value={item.label}
    />
  </TouchableOpacity>
);
const styles = StyleSheet.create({
  wrapper: {
    rowGap: 5,
    alignItems: 'center',
    justifyContent: 'center',
  },
  image: {
    justifyContent: 'center',
    alignItems: 'center',
    paddingVertical: 8,
    paddingHorizontal: 8,
    borderRadius: 12,
    width: 36,
    height: 36,
    backgroundColor: colors.background.tertiary,
  },
  imageActive: {
    backgroundColor: colors.background.secondary,
  },
});

export default memo(CategoryItem);
