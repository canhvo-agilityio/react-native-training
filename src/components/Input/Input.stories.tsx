import type { Meta, StoryObj } from '@storybook/react';
import Input from '.';

const meta = {
  title: 'components/Input',
  component: Input,
} satisfies Meta<typeof Input>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Flushed: Story = {
  args: {
    value: 't',
    onChangeText: () => {},
  },
};

export const Filled: Story = {
  args: {
    value: 'Enter your promo code',
    variant: 'filled',
    onChangeText: () => {},
  },
};

export const HasLabel: Story = {
  args: {
    value: '',
    onChangeText: () => {},
    label: 'Label',
  },
};

export const HasIcon: Story = {
  args: {
    value: '',
    onChangeText: () => {},
    label: 'Label',
  },
};
