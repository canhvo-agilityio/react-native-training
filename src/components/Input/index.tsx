import React, { memo } from 'react';
import { StyleSheet, TextInput, TouchableOpacity, View } from 'react-native';
import { InputVariantTypes } from '@/types';
import Text from '../Text';
import { colors, inputBaseStyle, inputVariants } from '@/themes';

export interface InputProps {
  value: string;
  placeholder?: string;
  label?: string;
  errorMessage?: string;
  variant?: InputVariantTypes;
  disabled?: boolean;
  icon?: React.JSX.Element;
  onChangeText: (text: string) => void;
  onPressIcon?: () => void;
}

const Input = ({
  value,
  placeholder,
  label,
  variant = 'flushed',
  errorMessage,
  disabled,
  icon,
  onChangeText,
  onPressIcon,
}: InputProps) => {
  return (
    <View style={[inputVariants[variant]]}>
      {label && (
        <Text
          value={label}
          variant="description"
          color={colors.text.quaternary}
        />
      )}
      <View style={styles.flex}>
        <TextInput
          value={value}
          placeholder={placeholder}
          editable={!disabled}
          onChangeText={onChangeText}
          style={[inputBaseStyle]}
        />
        {icon && onPressIcon ? (
          <TouchableOpacity testID="icon" onPress={onPressIcon}>
            {icon}
          </TouchableOpacity>
        ) : (
          icon
        )}
      </View>
      {errorMessage && <Text value={errorMessage} />}
    </View>
  );
};

export default memo(Input);

const styles = StyleSheet.create({
  flex: {
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingRight: 27,
    alignItems: 'center',
    gap: 14,
  },
});
