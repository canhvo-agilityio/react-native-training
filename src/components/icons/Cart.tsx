import Svg, { Path } from 'react-native-svg';

import { IconProps } from '@/types';
import { TouchableOpacity } from 'react-native';

export const Cart = ({
  width = 24,
  height = 25,
  color = '#808080',
  onPress,
}: IconProps) => (
  <TouchableOpacity testID="cart-icon" activeOpacity={0.8} onPress={onPress}>
    <Svg width={width} height={height} viewBox="0 0 24 25" fill="none">
      <Path
        fillRule="evenodd"
        clipRule="evenodd"
        d="M2.607 4.643A.643.643 0 013.25 4h1.928a.643.643 0 01.624.487l.52 2.084H21.25a.643.643 0 01.623.799l-1.928 7.714a.643.643 0 01-.624.487H7.75a.643.643 0 01-.624-.487l-2.45-9.798H3.25a.643.643 0 01-.643-.643zm4.037 3.214l1.607 6.429H18.82l1.607-6.429H6.644zm2.392 10.286a1.286 1.286 0 100 2.571 1.286 1.286 0 000-2.571zm-2.572 1.286a2.571 2.571 0 115.143 0 2.571 2.571 0 01-5.143 0zm11.572-1.286a1.286 1.286 0 100 2.571 1.286 1.286 0 000-2.571zm-2.572 1.286a2.571 2.571 0 115.143 0 2.571 2.571 0 01-5.143 0z"
        fill={color}
      />
    </Svg>
  </TouchableOpacity>
);
