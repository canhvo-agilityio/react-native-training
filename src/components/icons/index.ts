export * from './Show';
export * from './Hidden';
