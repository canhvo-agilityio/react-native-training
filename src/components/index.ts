export { default as Button } from './Button';
export { default as Text } from './Text';
export { default as Input } from './Input';
export { default as ProductCard } from './ProductCard';
export { default as ProductList } from './ProductList';
export { default as CategoryItem } from './CategoryItem';
export { default as Quantity } from './Quantity';
export { default as CartItem } from './CartItem';
