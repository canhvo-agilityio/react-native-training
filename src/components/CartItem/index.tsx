import { CartItem as CartItemType } from '@/types';
import { Image, StyleSheet, View } from 'react-native';
import Text from '../Text';
import Quantity from '../Quantity';
import { memo } from 'react';
import { TimesCircle } from '../icons';
import { colors, fontFamilies } from '@/themes';

interface CartItemProps {
  item: CartItemType;
}

const CartItem = ({ item }: CartItemProps) => {
  const { image, price, quantity, name } = item || {};
  return (
    <View style={styles.container}>
      <View style={styles.info}>
        <Image source={{ uri: image }} style={styles.image} />
        <View style={styles.detail}>
          <Text value={name} style={styles.name} />
          <Text value={`$ ${price}`} size="xmd" style={styles.price} />
          <Quantity quantity={quantity} />
        </View>
      </View>
      <TimesCircle />
    </View>
  );
};

export default memo(CartItem);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  image: {
    width: 100,
    height: 100,
    borderRadius: 10,
  },
  info: {
    flexDirection: 'row',
    gap: 20,
  },
  detail: {
    justifyContent: 'space-between',
  },
  name: {
    color: colors.border.quinary,
    fontFamily: fontFamilies.NunitoSansSemiBold,
  },
  price: {
    fontFamily: fontFamilies.NunitoSansBold,
    color: colors.primary,
  },
});
