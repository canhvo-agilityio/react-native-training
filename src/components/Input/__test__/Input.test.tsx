import { render, userEvent, screen } from 'test-utils';
import Input from '..';
import { Show } from '@/components/icons';

const handleChange = jest.fn();
const handlePressIcon = jest.fn();

describe('Input', () => {
  const mockProps = {
    value: '',
    placeholder: 'Enter text...',
    label: 'Username',
    onChangeText: handleChange,
  };

  it('should render properly', () => {
    const { toJSON } = render(<Input {...mockProps} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onChangeText callback when text changes', async () => {
    render(<Input {...mockProps} />);
    const input = screen.getByPlaceholderText('Enter text...');
    await userEvent.type(input, 'hello');
    expect(handleChange).toHaveBeenCalledTimes(5);
  });

  it('renders icon within TouchableOpacity if onPressIcon is provided', async () => {
    const mockPropsWithIcon = {
      ...mockProps,
      icon: <Show />,
      onPressIcon: handlePressIcon,
    };
    render(<Input disabled {...mockPropsWithIcon} />);
    const icon = screen.getByTestId('icon');
    await userEvent.press(icon);
    expect(handlePressIcon).toHaveBeenCalled();
  });

  it('renders errorMessage when errorMessage prop is provided', () => {
    const mockPropsWithError = {
      ...mockProps,
      errorMessage: 'Invalid input',
    };

    render(<Input {...mockPropsWithError} />);
    const errorMessageElement = screen.getByText('Invalid input');

    expect(errorMessageElement).toBeDefined();
  });
});
