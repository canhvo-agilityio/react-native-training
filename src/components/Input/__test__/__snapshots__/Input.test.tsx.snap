// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should render properly 1`] = `
<View
  style={
    [
      {
        "borderBottomColor": "#E0E0E0",
        "borderBottomWidth": 2,
        "color": "#909090",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#808080",
          "fontFamily": "NunitoSans-Regular",
        },
        {
          "fontSize": 14,
        },
        undefined,
        {
          "color": "#909090",
        },
      ]
    }
  >
    Username
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 14,
        "justifyContent": "space-between",
        "paddingRight": 27,
        "width": "100%",
      }
    }
  >
    <TextInput
      editable={true}
      onChangeText={[MockFunction]}
      placeholder="Enter text..."
      style={
        [
          {
            "color": "#909090",
            "fontFamily": "NunitoSans-Regular",
            "fontSize": 16,
            "fontWeight": "400",
            "paddingVertical": 11,
          },
        ]
      }
      value=""
    />
  </View>
</View>
`;
